### PR TITLE
RC: VBR (Mode 2) and CVBR (Mode 3) Rebase

### PR DESCRIPTION
**Pull request by anaghdin**
VBR (Mode 2) and CVBR (Mode 3) integrated from SVT-VP9

### CODE REVIEW

**Comment by j-mas** [comment]
Now encoding with rc 3 always stopp round about frame 1940/60 !!!
Encoding with the SvtAv1EncApp AppVeyor Version 1.0.378!
The HierarchicalLevels is always 3, not 4. 

C:\>SVT-AV1\SvtAv1EncApp.exe -lp 28 -w 1920 -h 1080 -enc-mode 8 -fps 25 -rc 3 -tbr 2500000 -intra-period 31 -irefresh-type 1 -lad 17 -scd 0 -sharp 0 -i "D:\Film1\1.yuv" -b "D:\Film1\1.ivf"
-------------------------------------------
SVT-AV1 Encoder
SVT [version]:  SVT-AV1 Encoder Lib v0.4.0
SVT [build]  :  Visual Studio 2017 64 bit
LIB Build date: Mar 25 2019 18:00:54
-------------------------------------------
Number of logical cores available: 28
Number of PPCS 67
-------------------------------------------
SVT [config]: Main Profile      Tier (auto)     Level (auto)
SVT [config]: EncoderMode                                                       : 8
SVT [config]: EncoderBitDepth / CompressedTenBitFormat                          : 8 / 0
SVT [config]: SourceWidth / SourceHeight                                        : 1920 / 1080
SVT [config]: FrameRate / Gop Size                                              : 25 / 32
SVT [config]: HierarchicalLevels / BaseLayerSwitchMode / PredStructure          : 3 / 0 / 2
SVT [config]: RCMode / TargetBitrate / LookaheadDistance / SceneChange          : Constraint VBR / 2500000 / 17 / 0
-------------------------------------------
Encoding      1955
C:\>SVT-AV1\SvtAv1EncApp.exe -lp 28 -w 1920 -h 1080 -enc-mode 8 -fps 25 -rc 3 -tbr 2500000 -intra-period 31 -irefresh-type 1 -lad 17 -scd 0 -sharp 0 -i "D:\Film1\1.yuv" -b "D:\Film1\1.ivf"
-------------------------------------------
SVT-AV1 Encoder
SVT [version]:  SVT-AV1 Encoder Lib v0.4.0
SVT [build]  :  Visual Studio 2017 64 bit
LIB Build date: Mar 25 2019 18:00:54
-------------------------------------------
Number of logical cores available: 28
Number of PPCS 67
-------------------------------------------
SVT [config]: Main Profile      Tier (auto)     Level (auto)
SVT [config]: EncoderMode                                                       : 8
SVT [config]: EncoderBitDepth / CompressedTenBitFormat                          : 8 / 0
SVT [config]: SourceWidth / SourceHeight                                        : 1920 / 1080
SVT [config]: FrameRate / Gop Size                                              : 25 / 32
SVT [config]: HierarchicalLevels / BaseLayerSwitchMode / PredStructure          : 3 / 0 / 2
SVT [config]: RCMode / TargetBitrate / LookaheadDistance / SceneChange          : Constraint VBR / 2500000 / 17 / 0
-------------------------------------------
Encoding      1950
C:\>SVT-AV1\SvtAv1EncApp.exe -lp 28 -w 1920 -h 1080 -enc-mode 8 -fps 25 -rc 3 -tbr 2500000 -intra-period 31 -irefresh-type 1 -lad 17 -scd 0 -sharp 0 -i "D:\Film1\1.yuv" -b "D:\Film1\1.ivf"
-------------------------------------------
SVT-AV1 Encoder
SVT [version]:  SVT-AV1 Encoder Lib v0.4.0
SVT [build]  :  Visual Studio 2017 64 bit
LIB Build date: Mar 25 2019 18:00:54
-------------------------------------------
Number of logical cores available: 28
Number of PPCS 67
-------------------------------------------
SVT [config]: Main Profile      Tier (auto)     Level (auto)
SVT [config]: EncoderMode                                                       : 8
SVT [config]: EncoderBitDepth / CompressedTenBitFormat                          : 8 / 0
SVT [config]: SourceWidth / SourceHeight                                        : 1920 / 1080
SVT [config]: FrameRate / Gop Size                                              : 25 / 32
SVT [config]: HierarchicalLevels / BaseLayerSwitchMode / PredStructure          : 3 / 0 / 2
SVT [config]: RCMode / TargetBitrate / LookaheadDistance / SceneChange          : Constraint VBR / 2500000 / 17 / 0
-------------------------------------------
Encoding      1949

**Comment by anaghdin** [comment]
At this point RC Mode 2 and 3 have some limitations. LAD must be equal to IntraPeriod. it is better not to specify LAD and let the encoder assign it automatically.

File: Source/Lib/Common/Codec/EbRateControlProcess.c [hunk 1]
```diff
@@ -494,422 +590,2772 @@ void HighLevelRcInputPictureMode2(
...
+                target_bit_rate = (uint64_t)((int64_t)parentpicture_control_set_ptr->target_bit_rate -
+                    MIN((int64_t)parentpicture_control_set_ptr->target_bit_rate * 3 / 4, (int64_t)(parentpicture_control_set_ptr->total_num_bits*context_ptr->frame_rate / (sequence_control_set_ptr->static_config.intra_period_length + 1)) >> RC_PRECISION))
+                    *rate_percentage_layer_array[sequence_control_set_ptr->static_config.hierarchical_levels][temporal_layer_idex] / 100;
+
+#if ADAPTIVE_PERCENTAGE
+                if (sequence_control_set_ptr->static_config.look_ahead_distance != 0 && sequence_control_set_ptr->intra_period_length != -1) {
+                    target_bit_rate = (uint64_t)((int64_t)parentpicture_control_set_ptr->target_bit_rate -
+                        MIN((int64_t)parentpicture_control_set_ptr->target_bit_rate * 3 / 4, (int64_t)(parentpicture_control_set_ptr->total_num_bits*context_ptr->frame_rate / (sequence_control_set_ptr->static_config.intra_period_length + 1)) >> RC_PRECISION))
+                        *parentpicture_control_set_ptr->bits_per_sw_per_layer[temporal_layer_idex] / sum_bits_per_sw;
+
```
Comment (mafonsoN) [inline Source/Lib/Common/Codec/EbRateControlProcess.c]: several spaces here

Comment (mafonsoN) [inline Source/Lib/Common/Codec/EbRateControlProcess.c]: commented code

**Comment by hassount** [comment]
@anaghdin 
Can you please fix the linux warning?

File: Source/Lib/Encoder/Codec/EbEncHandle.c [hunk 2]
```diff
@@ -440,7 +440,9 @@ int32_t set_parent_pcs(EbSvtAv1EncConfiguration*   config) {
         fps        = fps < 24  ? 24    : fps; 
         ppcs_count = MAX(min_ppcs_count, fps);
         ppcs_count = ((ppcs_count * 4) >> 1);  // 2 sec worth of internal buffering
-    
+#if RC
```
Comment (hassount) [inline Source/Lib/Encoder/Codec/EbEncHandle.c]: This is redundant as the fix was part of the previous commit.

File: Source/Lib/Common/Codec/EbDefinitions.h [hunk 3]
```diff
@@ -227,6 +227,11 @@ extern "C" {
 #define SC_HME_ME  0//use sc detector for hme/me setting
 
 #define AOM_INTERP_EXTEND 4
+#define RC                                              1 // VBR Rate control integrated from SVT-VP9
```
[the remark this answers is not in the document]
Comment (anaghdin) [reply Source/Lib/Common/Codec/EbDefinitions.h]: The macro is used to be able to keep track of the changes easily and will be removed later. RateControlMode  is still used as the input parameter

File: Source/App/EncApp/EbAppConfig.c [hunk 4]
```diff
@@ -407,7 +407,11 @@ void EbConfigCtor(EbConfig_t *config_ptr)
     config_ptr->look_ahead_distance                  = (uint32_t)~0;
     config_ptr->targetBitRate                        = 7000000;
     config_ptr->max_qp_allowed                       = 63;
+#if 1 //RC
```
[the remark this answers is not in the document]
Comment (anaghdin) [reply Source/App/EncApp/EbAppConfig.c]: Yes, min and max qp can be updated as an input. These are the default (suggested) values.

**Comment by EwoutH** [comment]
@anaghdin @hassount Can the [svt-av1_encoder_user_guide.md](https://github.com/OpenVisualCloud/SVT-AV1/blob/master/Docs/svt-av1_encoder_user_guide.md) be updated with the new rate control modes?

**Comment by jsunintel** [comment]
@anaghdin @hassount What is the RC2 mode? RC1 and RC2 are both named VBR, but what's the difference? How should I name those when adding such interface in ffmpeg plugin?

**Pull request merged**